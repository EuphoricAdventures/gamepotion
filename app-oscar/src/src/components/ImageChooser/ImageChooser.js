import React, { Component } from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'

import Heading2 from '../Heading2/Heading2'
import Input from '../Input/Input'
import Image from '../Image/Image'

import { get, set } from '../../localStorage'
import { font, colours } from '../../styleAbstractions'

const StyledImageChooser = styled.div`
  position: relative;
  // background-color: yellow;
  > .component--heading2 {
    height: 3rem;
    line-height: 3rem;
  }
  > .component--input {
    position: absolute;
    top: 0.5rem;
    right: 0;
    width: 12rem;
    // background-color: blue;
  }
  > div.outer {
    border: 1px solid #dadfe1;
    border-radius: 4px;
    overflow-y: scroll;
    > div.inner {
      height: 11rem;
      padding: 1rem 1rem 0 1rem;
      > div {
        position: relative;
        float: left;
        width: 8rem;
        height: 8rem;
        border-radius: 4px;
        // background-color: red;
        span {
          display: block;
          position: absolute;
          left: 0;
          top: 8rem;
          text-align: center;
          width: 8rem;
          height: 2rem;
          padding-top: 0.25rem;
          line-height: 0.75rem;
          ${font}
          font-size: 80%;
          color: ${colours.fore};
          cursor: default;
        }
      }
      > div + div {
        margin-left: 1rem;
      }
      > div.selected {
        background-color: #dadfe1;
      }
    }
  }
`

class ImageChooser extends Component {
  constructor(props) {
    super(props)
    let currentImage = this.props.currentImage || get(`component--image-chooser-current-image-${props.id}`)
    if (!currentImage && props.images.length > 0) {
      currentImage = props.images[0].id
      set(`component--image-chooser-current-image-${props.id}`, currentImage)
    }
    this.state = {
      currentImage,
      filter: get(`component--image-chooser-filter-${props.id}`) || ''
    }
    this.props.onChoose(currentImage)
    this.onUpdateFilter = this.onUpdateFilter.bind(this)
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    if (nextProps.currentImage !== this.props.currentImage) {
      this.setState({
        currentImage: nextProps.currentImage
      })
    }
  }

  onChoose(currentImage) {
    set(`component--image-chooser-current-image-${this.props.id}`, currentImage)
    this.props.onChoose(currentImage)
    this.setState({
      currentImage
    })
  }

  onUpdateFilter(filter) {
    set(`component--image-chooser-filter-${this.props.id}`, filter)
    this.setState({
      filter
    })
  }

  render() {
    const { filter } = this.state
    const images = (() => {
      if (filter === '') {
        return this.props.images
      }
      return this.props.images.filter(i => {
        return (
          i.name.toLowerCase().indexOf(filter.toLowerCase()) >= 0
        )
      })
    })()

    return (
      <StyledImageChooser className='component--image-chooser'>
        {this.props.title.length > 0 && <Heading2>{this.props.title}</Heading2>}
        <Input value={filter} onChange={this.onUpdateFilter} placeholder='Filter...' />
        <div className='outer'>
          <div className='inner' style={{width: `${(images.length * 144) - 16}px`}}>
            {images.map(i => {
              return (
                <div key={i.id} className={this.state.currentImage === i.id ? 'selected' : ''} onClick={() => this.onChoose(i.id)}>
                  <Image src={i.url} alt={i.name} />
                  <span>{i.name}</span>
                </div>
              )
            })}
          </div>
        </div>
      </StyledImageChooser>
    )
  }
}

ImageChooser.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string,
  images: PropTypes.array.isRequired,
  currentImage: PropTypes.string,
  onChoose: PropTypes.func
}

ImageChooser.defaultProps = {
  title: '',
  onChoose: () => {}
}

export default ImageChooser
