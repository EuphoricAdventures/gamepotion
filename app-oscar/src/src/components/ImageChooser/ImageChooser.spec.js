import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import ImageChooser from './ImageChooser'

jest.mock('uuid', () => jest.fn().mockReturnValue('id'))

it('renders', () => {
  const images = [
    {
      id: 'xyz',
      name: 'Xyz',
      url: 'xyz'
    }
  ]
  const tree = renderer
    .create(
      <ImageChooser id='test' images={images} currentImage='xyz' />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
