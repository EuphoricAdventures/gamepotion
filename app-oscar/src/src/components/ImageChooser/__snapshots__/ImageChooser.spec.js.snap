// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.c1 > img {
  display: block;
  max-width: calc(100% - 1rem);
  max-height: calc(100% - 1rem);
}

.c1 > p {
  width: 100%;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif;
  font-weight: 400;
  -webkit-text-size-adjust: 100%;
  text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;
  color: #2e3131;
  opacity: 0.5;
  position: absolute;
  top: 50%;
  text-align: center;
  -webkit-transform: translate(0,-50%);
  -ms-transform: translate(0,-50%);
  transform: translate(0,-50%);
}

.c1 > img {
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.c0 > .component--heading2 {
  margin-bottom: 1rem;
}

.c0 > div {
  border: 1px solid #dadfe1;
  border-radius: 4px;
  overflow-y: scroll;
}

.c0 > div > div.inner {
  height: 10rem;
  padding: 1rem;
}

.c0 > div > div.inner > div {
  position: relative;
  float: left;
  width: 8rem;
  height: 8rem;
  border-radius: 4px;
}

.c0 > div > div.inner > div span {
  display: block;
  position: absolute;
  left: 0;
  top: 8rem;
  text-align: center;
  width: 8rem;
  height: 2rem;
  line-height: 2rem;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif;
  font-weight: 400;
  -webkit-text-size-adjust: 100%;
  text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;
  font-size: 80%;
  color: #2e3131;
  cursor: default;
}

.c0 > div > div.inner > div + div {
  margin-left: 1rem;
}

.c0 > div > div.inner > div.selected {
  background-color: #dadfe1;
}

<div
  className="c0"
>
  <div>
    <div
      className="inner"
      style={
        Object {
          "width": "128px",
        }
      }
    >
      <div
        className="selected"
        onClick={[Function]}
      >
        <div
          className="component--image c1"
        />
        <span>
          Xyz
        </span>
      </div>
    </div>
  </div>
</div>
`;
